### PR TITLE
[13.0][ADD] account_move_line_cumulated_balance: Show field cumulated balance

### PR DESCRIPTION
**Pull request by cesar-tecnativa**
@Tecnativa TT37842

### CODE REVIEW

**Comment by cesar-tecnativa** [comment]
@pedrobaeza could you please review?

**Comment by cesar-tecnativa** [comment]
@chienandalu Please, could you review?

**Comment by cesar-tecnativa** [comment]
@victoralmau could you please review this?

**Comment by cesar-tecnativa** [comment]
@LudLaf could you review this?

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@LudLaf please review it functionally

**Comment by LudLaf** [comment]
> @LudLaf please review it functionally

Just tested. Functionality is working.

**Comment by pedrobaeza** [comment]
@LudLaf do the proper review with GitHub tools

**Review by LudLaf — COMMENTED** [review]
@cesar-tecnativa something has changed. When trying to test this module, it shows the following:

Error:


Traceback:
Error: fieldsViews is undefined
_generateActionViews/<@http://oca-account-financial-tools-13-0-pr1438-d1a1c2932c24.runboat.odoo-community.org/web/static/src/js/chrome/action_manager_act_window.js:355:30
_.forEach@http://oca-account-financial-tools-13-0-pr1438-d1a1c2932c24.runboat.odoo-community.org/web/static/lib/underscore/underscore.js:145:17
_generateActionViews@http://oca-account-financial-tools-13-0-pr1438-d1a1c2932c24.runboat.odoo-community.org/web/static/src/js/chrome/action_manager_act_window.js:353:11
_executeWindowAction/<@http://oca-account-financial-tools-13-0-pr1438-d1a1c2932c24.runboat.odoo-community.org/web/static/src/js/chrome/action_manager_act_window.js:258:30

**Comment by chienandalu** [comment]
> @cesar-tecnativa something has changed. When trying to test this module, it shows the following:


Please provide steps to reproduce the issue.

**Comment by LudLaf** [comment]
> > @cesar-tecnativa something has changed. When trying to test this module, it shows the following:
> 
> Please provide steps to reproduce the issue.
Peek 2022-08-11 10-33

**Comment by cesar-tecnativa** [comment]
> > > @cesar-tecnativa something has changed. When trying to test this module, it shows the following:
> > 
> > 
> > Please provide steps to reproduce the issue.
> > Peek 2022-08-11 10-33
> >       
> >         [
> >           
> >             Peek 2022-08-11 10-33
> >           
> >         ](https://user-images.githubusercontent.com/4016592/184095484-e1d7dc9f-5717-4caf-9d55-7038e913005d.gif)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://user-images.githubusercontent.com/4016592/184095484-e1d7dc9f-5717-4caf-9d55-7038e913005d.gif)
solved

**Review by LudLaf — COMMENTED** [review]
Module is shown as "To be installed", but doesn't allow you to keep installing nor using it

Peek 2022-08-11 12-29
.

**Comment by cesar-tecnativa** [comment]
new changes are done

**Comment by LudLaf** [comment]
> 



> new changes are done

Just tried to open the invoicing app, but didn't work.


---
Error:


Traceback:
Error: fieldsViews is undefined
_generateActionViews/<@http://oca-account-financial-tools-13-0-pr1438-82728466fadf.runboat.odoo-community.org/web/static/src/js/chrome/action_manager_act_window.js:355:30
_.forEach@http://oca-account-financial-tools-13-0-pr1438-82728466fadf.runboat.odoo-community.org/web/static/lib/underscore/underscore.js:145:17
_generateActionViews@http://oca-account-financial-tools-13-0-pr1438-82728466fadf.runboat.odoo-community.org/web/static/src/js/chrome/action_manager_act_window.js:353:11
_executeWindowAction/<@http://oca-account-financial-tools-13-0-pr1438-82728466fadf.runboat.odoo-community.org/web/static/src/js/chrome/action_manager_act_window.js:258:3
Peek 2022-08-15 10-22
0

**Review by LudLaf — CHANGES_REQUESTED** [review]
Not working on runboat nor when trying to install it in local, so it's not possible to test it functionally. Please check previous review.

File: account_move_line_cumulated_balance/models/account_move_line.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        currency_field="currency_id",
+        compute="_compute_cumulated_balance",
+        help="Cumulated balance in currency depending on the domain and "
+        "the order chosen in the view.",
+    )
+
+    @api.model
+    def search_read(self, domain=None, fields=None, offset=0, limit=None, order=None):
+        # Make an explicit order because we will need to reverse it
+        order = (order or self._order) + ", id"
```
Comment (chienandalu) [inline account_move_line_cumulated_balance/models/account_move_line.py]: Shouldn the order key go as something like `", id DESC"` or `", id ASC"`. What if `id` is already in the order string?

**Comment by chienandalu** [comment]
> Not working on runboat nor when trying to install it in local, so it's not possible to test it functionally. Please check previous review.

I just tested it on runbot. I had to install `account_lock_date_update` which was missing on some view for any reason.

File: account_move_line_cumulated_balance/models/account_move_line.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                    ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
+                ),
+                SUM(account_move_line.amount_currency)  OVER (
+                    ORDER BY {order_by_clause}
+                    ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
+                )
+            FROM account_move_line
+            LEFT JOIN account_move on account_move_line.move_id = account_move.id
+            WHERE
+            account_move.state = 'posted'
```
Comment (chienandalu) [inline account_move_line_cumulated_balance/models/account_move_line.py]: Why not filtering the `self` recordset `ids`. This query will be increasingly heavy, isn't it?

**Comment by pedrobaeza** [comment]
@chienandalu you can check now. Data is computed in a absolute way record per record with an aggregated SQL, but it's efficient enough on real databases, and it's the only way to get proper data no matter done filters.

**Comment by chienandalu** [comment]
> @chienandalu you can check now. Data is computed in a absolute way record per record with an aggregated SQL, but it's efficient enough on real databases, and it's the only way to get proper data no matter done filters.

And won't it produce tons of queries?

**Comment by pedrobaeza** [comment]
As you visualize a limited number of records in these views, that would be the number of queries. It can't be avoided due to the nature of the need.

**Review by chienandalu — APPROVED** [review]
:+1:

**Pull request merged**